sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -86,8 +87,12 @@ sealed trait Partitioning {
    */
   def satisfies(required: Distribution): Boolean
 
-  /** Returns the expressions that are used to key the partitioning. */
-  def keyExpressions: Seq[Expression]
+  /**
+   * Returns true iff we can say that the partitioning scheme of this [[Partitioning]]
+   * guarantees the same partitioning scheme described by `other`.
+   */
+  // TODO: Add an example once we have the `nullSafe` concept.
+  def guarantees(other: Partitioning): Boolean
```

Comment:
Do you think `semanticEqual` is a better name? I think this method is basically doing a equality check. For example, if `other` is a `HashPartitioning('a :: Nil, 10)` and `this` is a `SinglePartition`. We probably do not want to return true because the parent of this operator can be a join and the sibling of this operator can be HashPartitioned.